### PR TITLE
feat: Scheduling and Loading of the Mapping Table(RDCC-1867)

### PR DESCRIPTION
**Pull request by abhijitdiwan**
JIRA link (if applicable)
https://tools.hmcts.net/jira/browse/RDCC-1867

Change description
feat: Scheduling and Loading of the Mapping Table(RDCC-1867)

Does this PR introduce a breaking change? (check one with "x")
No

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -103,28 +104,52 @@ pmd {
   toolVersion = "6.18.0"
   sourceSets = [sourceSets.main, sourceSets.test, sourceSets.functionalTest, sourceSets.integrationTest, sourceSets.smokeTest]
   reportsDir = file("$project.buildDir/reports/pmd")
+  ignoreFailures = true
   // https://github.com/pmd/pmd/issues/876
-  ruleSets = []
-  ruleSetFiles = files("config/pmd/ruleset.xml")
+  //ruleSets = []
```
Comment (shreedhar-hmcts) [inline build.gradle:109]: remove comments if not required

File: .travis.yml [hunk 2]
```diff
@@ -25,5 +25,5 @@ script:
 
 after_success:
   - ./gradlew jacocoTestReport
-  - bash <(curl -s https://codecov.io/bash)
+#  - bash <(curl -s https://codecov.io/bash)
 #  - test "$TRAVIS_BRANCH" = "master" && test "$TRAVIS_PULL_REQUEST" = "false" && sh ./publish-swagger-docs.sh
```
Comment (shreedhar-hmcts) [inline .travis.yml:29]: remove comments if not required

File: src/functionalTest/resources/application-integration.yml [hunk 3]
```diff
@@ -0,0 +1,83 @@
+spring:
+  mail:
+    from: no-reply@reform.hmcts.net
+    to:  sushant.choudhari@hmcts.net,abhijit.diwan@hmcts.net,mallikarjun.puttana@hmcts.net
```
Comment (shreedhar-hmcts) [inline src/functionalTest/resources/application-integration.yml:4]: check the intended recipients

File: src/functionalTest/resources/application-integration.yml [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    subject:  JRD - Exception in processing -
```
Comment (shreedhar-hmcts) [inline src/functionalTest/resources/application-integration.yml:5]: subject needs to be changed to LRD ?

File: src/functionalTest/resources/application-integration.yml [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    # set batch false in integration test and run batch manually with batch runner to set dynamic test property values
+    # eg. dynamic properties ${parent.file.name}
+    job:
+      enabled: false
+
+#Local integration testing these values should be actual
+#ACCOUNT_NAME: ${ACCOUNT_NAME}
+#ACCOUNT_KEY: ${ACCOUNT_KEY}
+
+start-route: direct:LRD
```
Comment (shreedhar-hmcts) [inline src/functionalTest/resources/application-integration.yml:29]: can camel related properties be moved to another properties file ?

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/config/LrdCamelConfig.java [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+
+    // tasks configuration starts
+    @Bean
+    LrdRouteTask lrdRouteTask() {
+        return new LrdRouteTask();
+    }
+
+
+    @Bean
+    LrdExecutor jrdTask() {
```
Comment (shreedhar-hmcts) [inline src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/config/LrdCamelConfig.java:173]: rename to lrd

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/mapper/ServiceToCcdServiceMapper.java [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package uk.gov.hmcts.reform.locationrefdata.camel.mapper;
+
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.reform.data.ingestion.camel.mapper.IMapper;
+import uk.gov.hmcts.reform.locationrefdata.camel.binder.ServiceToCcdService;
+
+import java.util.HashMap;
```
Comment (KOTLAPRA23) [inline src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/mapper/ServiceToCcdServiceMapper.java:7]: are we following different rule set for LRD? why it  is not failing if imports are not in the proper order?

File: src/main/resources/Service.csv [hunk 8]
```diff
@@ -0,0 +1,2 @@
+service_code,ccd_service_name,ccd_jurisdiction_name
```
Comment (KOTLAPRA23) [inline src/main/resources/Service.csv:1]: is this file required in the src/main or test folder?

File: src/functionalTest/resources/application-integration.yml (hunk 5 above)
Comment (abhijitdiwan) [reply src/functionalTest/resources/application-integration.yml:29]: This is integartoin test properties..hence that is fine not source properties

**Review by KOTLAPRA23 — COMMENTED** [review]
please take care review comments

File: .travis.yml (hunk 2 above)
Comment (abhijitdiwan) [reply .travis.yml:29]: lets leave it it might need in future

File: build.gradle (hunk 1 above)
Comment (abhijitdiwan) [reply build.gradle:109]: done

File: src/functionalTest/resources/application-integration.yml (hunk 3 above)
Comment (abhijitdiwan) [reply src/functionalTest/resources/application-integration.yml:4]: done

File: Jenkinsfile_nightly [hunk 9]
```diff
@@ -6,9 +6,24 @@ properties([
...
 
 @Library("Infrastructure")
+import uk.gov.hmcts.contino.GradleBuilder
 
 def type = "java"
 def product = "rd"
 def component = "location-ref-data-load"
 
-withNightlyPipeline(type, product, component) {}
+GradleBuilder builder = new GradleBuilder(this, product)
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply Jenkinsfile_nightly:36]: done

File: build.gradle [hunk 10]
```diff
@@ -103,28 +104,52 @@ pmd {
...
+  //ruleSetFiles = files("config/pmd/ruleset.xml")
+  ruleSets = [
+    'category/java/errorprone.xml',
+    'category/java/bestpractices.xml'
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply build.gradle:110]: It is default ruleset comes as part of HMCTS..rather we have to change rd which is big work. We are not following right one in rd

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+        testInsertion();
+        jdbcTemplate.execute("delete from DATALOAD_SCHEDULAR_AUDIT");
+        integrationTestSupport.uploadFile(
+            "service-test.csv",
+            new FileInputStream(getFile(
+                "classpath:sourceFiles/service-test-update.csv"))
+        );
+
+        //camelContext.removeRouteDefinitions(camelContext.getRouteDefinitions());
```
Comment (AdilOoze) [inline src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:78]: is this needed?

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/mapper/ServiceToCcdServiceMapper.java (hunk 7 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/mapper/ServiceToCcdServiceMapper.java:7]: It is default ruleset comes as part of HMCTS..rather we have to change rd which is big work. We are not following right one in rd...Same as above.

File: src/main/resources/Service.csv (hunk 8 above)
Comment (abhijitdiwan) [reply src/main/resources/Service.csv:1]: right. removed

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 11 above)
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java]: rmoved

**Review by shreedhar-hmcts — DISMISSED** [review]

**Review by AdilOoze — APPROVED** [review]

File: src/main/resources/Service.csv [hunk 12]
```diff
@@ -0,0 +1,2 @@
...
+AA1,"service1,serive2",test-jurisdiction
```
Comment (shashank-rastogi) [inline src/main/resources/Service.csv:2]: file is wrong

there are 3 columns in file. - 
1.  SERVICE ID 
2.  CCD Jurisdiction Name 
3. CCD CASE TYPE

the third column which is CCD case type is a comma separated value and not the second one

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+     * @param serviceToCcdServices serviceToCcdServices
+     * @return List
+     */
+    private List<ServiceToCcdService> populateService(List<ServiceToCcdService> serviceToCcdServices) {
+        List<ServiceToCcdService> refinedServiceToCcdServices = new ArrayList<>();
+        serviceToCcdServices.forEach(serviceToCcdService ->
+                                         of(serviceToCcdService.getCcdServiceName().split(","))
+                                             .forEach(serviceNames -> {
+                                                 refinedServiceToCcdServices.add(ServiceToCcdService.builder()
+                                                                                     .serviceCode(serviceToCcdService
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java:78]: this logic will change based on the column name comment mentioned on the file.
Column which can have comma separated values is CCD CASE TYPE

**Review by shashank-rastogi — CHANGES_REQUESTED** [review]

File: build.gradle [hunk 14]
```diff
@@ -212,7 +240,7 @@ dependencies {
   compile group: 'javax.el', name: 'javax.el-api', version: '3.0.0'
   compile group: 'org.flywaydb', name: 'flyway-core', version: versions.flyway
   compile group: 'org.postgresql', name: 'postgresql', version: '42.2.14'
-
+  compile group: 'uk.gov.hmcts.reform', name: 'data-ingestion-lib', version: '0.3.7-beta'
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply build.gradle:240]: This is not production till now...lets complete master testing and will rerelease.

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/binder/ServiceToCcdService.java [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    @DataField(pos = 1, columnName = "service_code")
+    @NotEmpty
+    String serviceCode;
+
+    @DataField(pos = 2, columnName = "ccd_service_name")
+    @NotEmpty
+    String ccdServiceName;
+
+    @DataField(pos = 3, columnName = "ccd_jurisdiction_name")
+    String ccdJurisdictionName;
```
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/binder/ServiceToCcdService.java:34]: Yes it is as per previous...db model last sprint. where have 3 columns  service_code, ccd_service_name, ccd_jurisdiction_name. Will refactor as current

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/config/LrdCamelConfig.java [hunk 16]
```diff
@@ -0,0 +1,222 @@
...
+    // db configuration starts
+    private static final PostgreSQLContainer testPostgres = new PostgreSQLContainer("postgres")
+        .withDatabaseName("dbjuddata_test");
+
+    static {
+        testPostgres.start();
+    }
+
+    @Bean
+    public DataSource dataSource() {
```
[the remark this answers is not in the document]
Comment (shashank-rastogi) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/config/LrdCamelConfig.java:113]: i dont see this being changed?

File: src/functionalTest/resources/application-integration.yml [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+invalid-jsr-sql: insert into dataload_exception_records(table_name,scheduler_start_time,scheduler_name,key,field_in_error,error_description,updated_timestamp) values(?, ?, ?, ?, ?, ?, ?)
+jsr-threshold-limit: 5
+batchjob-name: LocationRefDataLoad
+scheduler-insert-sql:  insert into dataload_schedular_audit (scheduler_name,scheduler_start_time,scheduler_end_time,scheduler_status)
+  values(?,?,?,?)
+audit-enable: true
+lrd-route-task: lrdRouteTask
+
+routes-to-execute-orchestration: lrd-service
+route:
```
[the remark this answers is not in the document]
Comment (shashank-rastogi) [reply src/functionalTest/resources/application-integration.yml:49]: ??

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/binder/ServiceToCcdCaseType.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+    @DataField(pos = 1, columnName = "service_code")
+    @NotEmpty
+    String serviceCode;
+
+    @DataField(pos = 2, columnName = "ccd_service_name")
+    String ccdServiceName;
+
+    @DataField(pos = 3, columnName = "ccd_case_type")
+    @NotEmpty
+    String ccdCaseType;
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/binder/ServiceToCcdCaseType.java:34]: ccd case type is not empty and there is possibility the the file may not have any value in this column for a row

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+        );
+
+        audit(serviceToCcdServiceJsrValidatorInitializer, exchange);
+
+        if (filteredServiceToCcdCaseTypes.size() == 0) {
+            log.error(" {} ServiceToCcdService failed as no valid records present::", logComponentName);
+
+            throw new RouteFailedException("ServiceToCcdService failed as no valid records present");
+        }
+        exchange.getMessage().setBody(filteredServiceToCcdCaseTypes);
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java:63]: looks like we are not handling scenarios where ccd case type can be null in file
There are three possibilities and all should be handled -

1. All columns have value in csv
2. Some of the rows dont have ccd case type. In this case ccd case type should be blank in table
3. Some rows dont have ccd service name and case type. In this case it should be ignored and no record should be inserted. If there is already a record for service code then it should be deleted from the table

File: src/functionalTest/resources/sourceFiles/service-test-failure.csv [hunk 20]
```diff
@@ -0,0 +1,2 @@
+service_code,ccd_service_name,ccd_case_type
+AAA2,ccd-service2,""
```
Comment (shashank-rastogi) [inline src/functionalTest/resources/sourceFiles/service-test-failure.csv:2]: why will this fail??

Comment (shashank-rastogi) [reply src/functionalTest/resources/sourceFiles/service-test-failure.csv:2]: file wont have double quotes in the end.

File: src/functionalTest/resources/sourceFiles/service-test.csv [hunk 21]
```diff
@@ -0,0 +1,3 @@
+service_code,ccd_service_name,ccd_case_type
+AAA1,ccd-service1,"service1,service2"
```
Comment (shashank-rastogi) [inline src/functionalTest/resources/sourceFiles/service-test.csv:2]: Valid file -

service_code,ccd_service_name,ccd_case_type
AAA1,ccd-service1,"service1,service2"
AAA2,ccd-service2,"service11,service12"
AAA3, ccd-service3,
AAA4,,

Comment (shashank-rastogi) [reply src/functionalTest/resources/sourceFiles/service-test.csv:2]: ideally file wont include - AAA4,, but we should be able to handle it and ignore in case its there

**Review by shashank-rastogi — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+    public void testTaskletSuccess() throws Exception {
+        testInsertion();
+    }
+
+    @Test
+    @Sql(scripts = {"/testData/truncate-lrd.sql"})
+    public void testTaskletSuccessWithUpdateAndDelete() throws Exception {
+
+        testInsertion();
+        jdbcTemplate.execute("delete from DATALOAD_SCHEDULAR_AUDIT");
```
Comment (shashank-rastogi) [inline src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:71]: i dont think we should execute this delete statement here

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+        integrationTestSupport.uploadFile(
+            "service-test.csv",
+            new FileInputStream(getFile(
+                "classpath:sourceFiles/service-test-update.csv"))
+        );
```
Comment (shashank-rastogi) [inline src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:76]: we should also include a scenario where we only have to delete and no insert

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java (hunk 13 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java]: done

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/binder/ServiceToCcdCaseType.java (hunk 18 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/binder/ServiceToCcdCaseType.java:33]: done

File: src/functionalTest/resources/sourceFiles/service-test-failure.csv (hunk 20 above)
Comment (abhijitdiwan) [reply src/functionalTest/resources/sourceFiles/service-test-failure.csv]: have a look updated file.

File: src/functionalTest/resources/sourceFiles/service-test.csv (hunk 21 above)
Comment (abhijitdiwan) [reply src/functionalTest/resources/sourceFiles/service-test.csv:2]: This is covered in unit test. Ideally code logic is handled in unit cases.

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 22 above)
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:71]: This is inser and update and we have idempotent job how you test day2 scenario? this is needed.

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 23 above)
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:76]: can you put use case may be in jira? I am not clear what should excepted here.

File: src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java (hunk 19 above)
Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/locationrefdata/camel/processor/ServiceToCcdServiceProcessor.java]: done

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 22 above)
Comment (shashank-rastogi) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:71]: You can do it in multiple ways.. You are only asserting by checking record in  audit table which you can do by doing a select with order by statement as well.

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 23 above)
Comment (shashank-rastogi) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:76]: It is already included in the acceptance criteria under #c of 2 and  #3

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/testsupport/IntegrationTestSupport.java [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Date;
+import javax.annotation.PostConstruct;
+import javax.annotation.Resource;
+
+import static com.microsoft.azure.storage.blob.DeleteSnapshotsOption.INCLUDE_SNAPSHOTS;
+
+@Component
+@ContextConfiguration(classes = {
+    AzureBlobConfig.class, StorageCredentials.class}, initializers = ConfigFileApplicationContextInitializer.class)
+public class IntegrationTestSupport {
```
[the remark this answers is not in the document]
Comment (shashank-rastogi) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/testsupport/IntegrationTestSupport.java:26]: ?

File: src/functionalTest/resources/sourceFiles/service-test-failure.csv (hunk 20 above)
Comment (shashank-rastogi) [reply src/functionalTest/resources/sourceFiles/service-test-failure.csv]: i dont see it as updated...

Comment (shashank-rastogi) [reply src/functionalTest/resources/sourceFiles/service-test-failure.csv]: And my previous question was why is this scenario expected to fail?
I think it is a valid scenario and should pass and not fail

File: src/functionalTest/resources/sourceFiles/service-test-partial-success.csv [hunk 25]
```diff
@@ -0,0 +1,3 @@
+service_code,ccd_service_name,ccd_case_type
+AAA1,ccd-service1,"service1,service2"
+AAA2,ccd-service2,""
```
Comment (shashank-rastogi) [inline src/functionalTest/resources/sourceFiles/service-test-partial-success.csv]: same as above

File: src/functionalTest/resources/sourceFiles/service-test.csv (hunk 21 above)
Comment (shashank-rastogi) [reply src/functionalTest/resources/sourceFiles/service-test.csv:2]: This is not about code logic. This should be tested in functional Include this scenario in functional test

**Review by shashank-rastogi — CHANGES_REQUESTED** [review]

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 22 above)
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:71]: What is wrong in this way?? I don't get that. delete the record in test won't harm

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 23 above)
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:76]: ok..This is just file name will rename this

File: src/functionalTest/resources/sourceFiles/service-test-failure.csv (hunk 20 above)
Comment (abhijitdiwan) [reply src/functionalTest/resources/sourceFiles/service-test-failure.csv]: We are failing in case file does not have any valid record. If all records are invalid.

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java (hunk 23 above)
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/LrdBatchApplicationTest.java:76]: service-test-empty-case-or-name created new file

File: src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/testsupport/IntegrationTestSupport.java (hunk 24 above)
[the remark this answers is not in the document]
Comment (abhijitdiwan) [reply src/functionalTest/java/uk/gov/hmcts/reform/locationrefdata/cameltest/testsupport/IntegrationTestSupport.java:26]: done

File: Jenkinsfile_CNP [hunk 26]
```diff
@@ -10,6 +10,8 @@ properties(
 )
 
 @Library("Infrastructure")
+import uk.gov.hmcts.contino.GradleBuilder
+
```
Comment (shashank-rastogi) [inline Jenkinsfile_CNP:14]: @abhijitdiwan you need to fix the sonar code smells

**Review by shashank-rastogi — CHANGES_REQUESTED** [review]

**Review by shashank-rastogi — DISMISSED** [review]

**Review by KOTLAPRA23 — APPROVED** [review]

**Review by shashank-rastogi — APPROVED** [review]
